lib/flight_scheduler/schedulers/base_scheduler.rb

```diff
@@ -150,20 +157,22 @@ def candidates(partition)
         end
 
         if job.job_type == 'ARRAY_JOB'
-          previous_task = nil
-          loop do
-            next_task = job.task_generator.next_task
-            if next_task == previous_task
-              # We failed to schedule the ARRAY_TASK.  Move onto the next
-              # ARRAY_JOB or JOB.
-              break
-            elsif next_task.nil?
-              # We've exhausted the ARRAY_JOB.  Move onto the next ARRAY_JOB
-              # or JOB.
-              break
-            else
-              previous_task = next_task
-              yielder << next_task
+          catch :done_with_array_job do
+            previous_task = nil
+            loop do
+              next_task = job.task_generator.next_task
+              if next_task == previous_task
+                # We failed to schedule the ARRAY_TASK.  Move onto the next
+                # ARRAY_JOB or JOB.
+                throw :done_with_array_job
```

Comment:
That's how I used to do it.  You complained that it was hard to follow due to the `loop`/`break` being nested inside of an each `each`.

Reply:
Ah fair point. My initial comment was more highlighting the complexity of the process flow rather than the exact syntax.

Keep the `catch`/`throw` for now as it works, eventually we may encapsulate the logic into another object.